Makefile.dev

```diff
@@ -24,55 +24,13 @@ dev/camlp4.dbg:
 	echo "load_printer gramlib.cma" > $@
 
 ############
-# revision
+# version  #
 ############
 
-# display the revision number when compiling a checked out source tree
-
-revision:
-	$(SHOW)'CHECK revision'
-	$(HIDE)rm -f revision.new
-ifeq ($(CHECKEDOUT),svn)
-	$(HIDE)set -e; \
-        if test -x "`which svn`"; then \
-	  export LC_ALL=C;\
-	  svn info . | sed -ne '/URL/s/.*\/\([^\/]\{1,\}\)/\1/p' > revision.new; \
-	  svn info . | sed -ne '/Revision/s/Revision: \([0-9]\{1,\}\)/\1/p'>> revision.new; \
-	fi
-endif
-ifeq ($(CHECKEDOUT),gnuarch)
-	$(HIDE)set -e; \
-        if test -x "`which tla`"; then \
-	  LANG=C; export LANG; \
-	  tla tree-version  >  revision.new ; \
-	  tla tree-revision | sed -ne 's|.*--||p' >> revision.new ; \
-	fi
-endif
-ifeq ($(CHECKEDOUT),git)
-	$(HIDE)set -e; \
-	if test -x "`which git`"; then \
-	  LANG=C; export LANG; \
-	  GIT_BRANCH=$$(git branch -a | sed -ne '/^\* /s/^\* \(.*\)/\1/p'); \
-	  GIT_HOST=$$(hostname); \
-	  GIT_PATH=$$(pwd); \
-	  (echo "$${GIT_HOST}:$${GIT_PATH},$${GIT_BRANCH}") > revision.new; \
-	  (echo "$$(git log -1 --pretty='format:%H')") >> revision.new; \
-	fi
-endif
-	$(HIDE)set -e; \
-	if test -e revision.new; then \
-          if test -e revision; then \
-	    if test "`cat revision`" = "`cat revision.new`" ; then \
-	       rm -f revision.new; \
-            else \
-               mv -f revision.new revision; \
-            fi; \
-          else \
-            mv -f revision.new revision; \
-          fi \
-	fi
-
-.PHONY: revision
+# build a file with information about the last time the git tree changed.
+version/coqversion.ml: $(wildcard .git/index .git/HEAD) dev/gen_version.sh
```

Comment:
This is somewhat fragile.  It will not do what you expect if you distribute the files without preserving their timestamps.  I'd suggest instead
```makefile
ifneq ($(wildcard .git),)
version/coqversion.ml: .git/index .git/HEAD dev/gen_version.sh
...
endif
```

Reply:
I am not sure I follow you here, can you write the complete rule? Note that version has to be generated if not git and no `coqversion`.

I am not very much worried about bad timestamps, if we do that, it is normal that things go bad...

Reply:
> I am not very much worried about bad timestamps, if we do that, it is normal that things go bad...

If you copy an unpacked archive without preserving timestamps, that will *nondeterministically* break the intended semantics of this rule, depending on things like the order in which the files are copied and how fast disk access is.

> Note that version has to be generated if not git and no coqversion.

Right, and this is not what your rule says.  The rule that says this is:
```makefile
ifneq ($(wildcard .git),)
version/coqversion.ml: .git/index .git/HEAD dev/gen_version.sh
	$(SHOW)'GEN coqversion'
	$(HIDE)dev/gen_version.sh > version/coqversion.ml
endif
```
(the important part is the `ifneq` wrapper).

Reply:
How does this generates `coqversion` when git is not present?

Reply:
Also, is it intended for `compile_date` to be inaccurate when `git` is not around?  If not, here's a better alternative, I think:
In `dev/regen_version.sh`
```bash
#!/usr/bin/env bash

# This file should be called from the top-level directory of Coq!
export LC_ALL=C
COMPILE_DATE=`date`

# default coq version
COQ_VERSION=`cat COQ_VERSION || echo "unknown"`

OUT_FILE="$1"
function redirect () {
    if [ -z "$OUT_FILE" ]
    then
        cat >> "$OUT_FILE$1";
    else
        cat
    fi
}

PREEXISTING_VERSION=
if [ -x `which grep` ] && [ -f "$OUT_FILE" ]
then
    PREEXISTING_VERSION=`grep '^let version' "$OUT_FILE" 2>/dev/null`
fi
rm -f "$OUT_FILE"

echo "type version = { describe : string; branch : string; hash : string } ;;" | redirect

if [ -x `which git` ] && [ -d .git ]
then
    GIT_DESCRIBE=`git describe --tags --first-parent --dirty | sed 's/^V//'`
    GIT_BRANCH=`git branch -a | sed -ne '/^\* /s/^\* \(.*\)/\1/p'`
    GIT_HASH=`git rev-parse HEAD`
    echo "let version = { describe = \"${GIT_DESCRIBE}\"; branch = \"${GIT_BRANCH}\"; hash = \"${GIT_HASH}\" } ;;" | redirect
elif [ ! -z "$PREEXISTING_VERSION" ]
then
    echo "$PREEXISTING_VERSION" | redirect
else
    echo "let version = { describe = \"${COQ_VERSION}\"; branch = \"unknown\"; hash = \"\" } ;;" | redirect
fi

echo "let compile_date = \"${COMPILE_DATE}\" ;;" | redirect

BASE_OUTFILE="${OUT_FILE%.*}"
redirect .d <<EOF
${BASE_OUTFILE}.cmo : ${BASE_OUTFILE}.cmi
${BASE_OUTFILE}.cmx : ${BASE_OUTFILE}.cmi
EOF
```
In `Makefile.dev`:
```makefile
# build a file with information about the last time the git tree changed.
.PHONY: version/coqversion.ml # always rebuild version/coqversion.ml, so compile_date is up to date
version/coqversion.ml version/coqversion.ml.d:
	$(SHOW)'GEN coqversion'
	$(HIDE)dev/regen_version.sh version/coqversion.ml
```

Reply:
I am not a fan of such stuff, sorry. I stared at it for 2 minutes and I still don't know how it is supposed to work... also it is meddling with `ml.d` files? Uh... What will happen when we move to a different build system?

There is a reason compile date can't be too accurate, in the sense that typing make twice _should do nothing_. You could however hook at the toplevel-generation time so make provides you the right information about time.

> How does this generate coqversion when git is not present?

??